internal/framework/generic_resource.go

```diff
@@ -333,46 +364,167 @@ func isNetworkError(err error) bool {
+	stream.lock.Lock()
+	defer stream.lock.Unlock()
+	return fn()
+}
+
+func sendToChannel[T any](ctx context.Context, ch chan T, v T) {
+	select {
+	case ch <- v:
+	case <-ctx.Done():
+		tflog.Debug(ctx, "Cancelling send because context is done")
+	}
+}
+
+func (r *GenericResource) stream(ctx context.Context, state ResourceState) *eventStream {
+	stream := eventStream{
+		eventChannel: make(chan bool),
+		errChannel:   make(chan error),
+	}
+	stream.wg.Add(1)
+	go func() {
+		defer stream.wg.Done()
+		// Try to use SSE to stream events on resource
+		err := r.client.ProviderConfig.ConsumeEvents(ctx, state.GetEventPath(), func(event sse.Event) {
+			// Do nothing on heartbeat messages
+			if event.Type == SSE_EVENT_HEARTBEAT {
+				return
+			}
+			// If event is DELETED or has no data, notify that the resource was deleted
+			if event.Type == SSE_EVENT_DELETED || event.Data == SSE_DATA_NO_RESOURCE {
+				sendToChannel(ctx, stream.eventChannel, false)
+			} else {
+				// Deserialize message and notify event
+				err := stream.withLock(func() error {
+					tflog.Debug(ctx, "Unmarshalling data from SSE message",
+						map[string]any{"event": event.Type, "data": event.Data})
+					return json.Unmarshal([]byte(event.Data), state)
+				})
+				if err == nil {
+					sendToChannel(ctx, stream.eventChannel, true)
+				} else {
+					sendToChannel(ctx, stream.errChannel, err)
+				}
+			}
+		})
+		// If context is not done, there was an error consuming SSE messages
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			tflog.Info(ctx, "Downgrading from SSE to polling", map[string]any{"error": err})
+		}
+		// Use polling to generate resource events
+		for {
+			err := stream.withLock(func() error {
+				return state.Read(ctx, r.client.Client)
+			})
+			if err == nil {
+				sendToChannel(ctx, stream.eventChannel, true)
+			} else if helper.IsNotFound(err) {
+				sendToChannel(ctx, stream.eventChannel, false)
+			} else if isNetworkError(err) {
+				// Suppress network errors, which may be transient and retriable
+				tflog.Info(ctx, "Suppressing network error while awaiting readiness",
+					map[string]any{"resourceType": r.TypeName, "error": err.Error()})
+			} else {
+				sendToChannel(ctx, stream.errChannel, err)
+			}
+			// Wait for polling interval or until context is done
+			select {
+			case <-time.After(POLLING_INTERVAL):
+				continue
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return &stream
+}
+
 func (r *GenericResource) AwaitReady(ctx context.Context, state ResourceState, operation string) error {
 	timeout := r.GetTimeout(operation, READINESS_TIMEOUT)
 	if timeout == 0 {
 		tflog.Info(ctx, "Not waiting for resource to achieve desired state because timeout is 0",
 			map[string]any{"resourceType": r.TypeName, "operation": operation})
 		return nil
 	}
-	consecutiveFailed := 0
 	ctx, cancel := context.WithTimeout(ctx, timeout)
-	defer cancel()
+	stream := r.stream(ctx, state)
+	defer func() {
+		cancel()
+		stream.wg.Wait()
```

Comment:
Is this supposed to enable the caller of `AwaitReady()` or `AwaitDeleted()` to cancel its context? Or are we trying to guard against the goroutine never invoking `stream.wg.Done()`?

If it is the latter, I can replace the `WaitGroup` with a channel and wait for a small timeout. But I think we really want the goroutine to exit, because nothing outside this function guards access to the state based on the lock, which means that we could have a rogue goroutine changing out the state from under us.

Reply:
Using `WaitGroup` is fine. What I was suggesting will prevent the following case --- If the goroutine doesn't end, `wg.Wait()` will block forever and will prevent the `main` gorotine from exiting.

Reply:
I think it might be preferable to block progress completely, because if we cannot guarantee that the SSE consumer is done then we have a rogue goroutine potentially invalidating data before we store it in the Terraform state.